File: worlds/thrift_agent_server.go
package worlds

import (
	"context"
	"fmt"

	"github.com/Astera-org/worlds/network"
	"github.com/Astera-org/worlds/network/gengo/env"
	log "github.com/zajann/easylog"
)

const ADDR = "localhost:9090"

type AgentHandler struct {
	env.Agent
}

func (agent AgentHandler) Init(ctx context.Context, actionSpace env.Space,
	observationSpace env.Space) (map[string]string, error) {
	return map[string]string{"name": "GoAgentZero"}, nil
}

func (agent AgentHandler) Step(ctx context.Context, observations env.Observations, debug string) (env.Actions, error) {
	log.Debug("Step called observations:", observations, "debug:", debug)
	action := env.Action{DiscreteOption: 0}
	return env.Actions{"move": &action}, nil
}

func main() {
	log.Init(
		log.SetLevel(log.INFO),
		log.SetFileName("thrift_agent_server.log"),
	)

	handler := AgentHandler{}
	server := network.MakeServer(handler)
	log.Info("listening on", ADDR)
	server.Serve()
}

File: worlds/network/thrift_agent_server.py
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol
from thrift.server import TServer

from network.genpy.env import Agent
from network.genpy.env.ttypes import Action, ETensor, Shape


HOST = '127.0.0.1'
PORT = '9090'


class AgentHandler:

    def __init__(self):
        self.name = "AgentZero"

    def init(self, actionSpace, observationSpace):
        print(f'init called actionsSpace: {actionSpace} observationSpace: {observationSpace}')
        self.moveActionSpace = actionSpace["move"]
        self.discreteLabels = self.moveActionSpace.discreteLabels
        print(f'discrete labels {self.discreteLabels}')
        self.observationSpace = observationSpace
        return {"name": self.name}

    def step(self, observations, debug):
        print(f'step called obs: {observations} debug: {debug}')
        if self.discreteLabels:
            action = Action(discreteOption=0)
        else:
            action = Action(vector=ETensor(shape=Shape(shape=[1]), values=[1.0]))
        return {"move": action}

def setup_server(handler):
    processor = Agent.Processor(handler)
    transport = TSocket.TServerSocket(host=HOST, port=PORT)
    tfactory = TTransport.TBufferedTransportFactory()
    pfactory = TBinaryProtocol.TBinaryProtocolFactory()
    server = TServer.TSimpleServer(processor, transport, tfactory, pfactory)
    print(f'listening on {HOST}:{PORT}')
    return server


if __name__ == '__main__':
    handler = AgentHandler()
    server = setup_server(handler)
    server.serve()
